### PR TITLE
[REFACTOR] project pipeline command

### PR DESCRIPTION
**Pull request by dennisseah**
1. Create a set of bedrockYaml helper functions so that 
   we have a common place for `bedrock.yaml` related functions
   a. e.g. isBedrockFileExists function is in create-variable-group.ts
      and it is consumed by create-variable-group-test.ts.
   b. create a dummy `bedrock.yaml` in a random temporay directory is 
      also a common used one
   c. reading `bedrock.yaml` as a JSON object is another one
   d. create a constant `YAML_NAME` for `bedrock.yaml`
   e. have a `addNewService` function
   will expand to have more features like write `bedrock.yaml`, etc.

2. Remove the validateRequiredArguments function in create-variable-group.ts
   the functionality is handled by `CommandBuilder.validateForRequiredValues`
   function

3. simplify pipeline.test.ts as mock values and objects are repeated many times

4. pipeline.ts
   a. move command decoration out to a JSON file
   b. declare an interface for the expected values from commandline
   c. replace `isValidConfig` function with `CommandBuilder.validateForRequiredValues`

5. miscellaneous changes.

will fix the status code issue once this PR is merged https://github.com/CatalystCode/spk/pull/214

### CODE REVIEW

**Comment by andrebriggs** [comment]
@dennisseah changes look good. Do you still need to do an up date because of #214?

**Comment by dennisseah** [comment]
> @dennisseah changes look good. Do you still need to do an up date because of #214?

you are right. I need to do it. Just did it; and tests are passing. Thanks

**Comment by andrebriggs** [comment]
Overall looks good to me. 

@mtarng @sarath-p @edaena would love your thoughts

**Review by andrebriggs — APPROVED** [review]

File: src/commands/project/create-variable-group.test.ts [hunk 1]
```diff
@@ -297,37 +204,19 @@ describe("setVariableGroupInBedrockFile", () => {
...
-      services: {},
-      variableGroups: []
-    };
-
-    const asYaml = yaml.safeDump(bedrockFileData, {
-      lineWidth: Number.MAX_SAFE_INTEGER
-    });
-
-    fs.writeFileSync(path.join(randomTmpDir, "bedrock.yaml"), asYaml);
+    const randomTmpDir = createBedrockYaml();
```
Comment (sarath-p) [inline src/commands/project/create-variable-group.test.ts:207]: does file gets created in randomTempDir? or Default dir?

File: src/commands/project/create-variable-group.test.ts [hunk 2]
```diff
@@ -297,37 +204,19 @@ describe("setVariableGroupInBedrockFile", () => {
...
 
     await setVariableGroupInBedrockFile(randomTmpDir, variableGroupName);
 
-    const bedrockFilePath = path.join(randomTmpDir, "bedrock.yaml");
-    expect(fs.existsSync(bedrockFilePath)).toBe(true);
-
-    const bedrockFile = readYaml<IBedrockFile>(bedrockFilePath);
+    expect(isBedrockFileExists(randomTmpDir)).toBe(true);
+    const bedrockFile = readBedrockFile(randomTmpDir);
```
Comment (sarath-p) [inline src/commands/project/create-variable-group.test.ts:212]: In relation to above comment, these two lines are expecting a file in random dir. May be randomDir needs to be passed.

**Review by sarath-p — APPROVED** [review]
Couple of comments.

File: src/commands/project/pipeline.ts [hunk 3]
```diff
@@ -18,252 +24,180 @@ import {
...
-export const installLifecyclePipeline = async (
-  orgName: string,
-  personalAccessToken: string,
-  pipelineName: string,
-  repositoryName: string,
-  repositoryUrl: string,
-  project: string,
-  buildScriptUrl: string,
-  exitFn: (status: number) => void
+export const execute = async (
```
Comment (mtarng) [inline src/commands/project/pipeline.ts:105]: Can this method name be more descriptive? ie. `executeInstallPipelineCommand`?

File: src/commands/project/pipeline.ts [hunk 4]
```diff
@@ -18,252 +24,180 @@ import {
...
+  });
+};
+
+const createPipeline = async (
+  values: ICommandOptions,
+  devopsClient: IBuildApi
+): Promise<BuildDefinition> => {
   const definition = definitionForAzureRepoPipeline({
-    branchFilters: ["master"],
+    branchFilters: ["master"], // TOFIX: hardcoded?
```
Comment (mtarng) [inline src/commands/project/pipeline.ts]: I believe this is intentional. Our pipelines are triggered only by merges into the master branch.

@evanlouie @bnookala

**Review by mtarng — CHANGES_REQUESTED** [review]

File: src/commands/project/create-variable-group.test.ts (hunk 1 above)
Comment (dennisseah) [reply src/commands/project/create-variable-group.test.ts:207]: in random temp dir.

File: src/commands/project/create-variable-group.test.ts (hunk 2 above)
Comment (dennisseah) [reply src/commands/project/create-variable-group.test.ts:212]: sorry I do not get it. Can you elaborate?

File: src/commands/project/pipeline.ts (hunk 3 above)
Comment (dennisseah) [reply src/commands/project/pipeline.ts:105]: the goal is the standard all command related file. that is it shall have a
1. commandDecorator function
2. execute function.

Once we standardize all these files, we can use reflective to get the required function and execute them accordingly.

File: src/commands/project/pipeline.ts (hunk 4 above)
Comment (dennisseah) [reply src/commands/project/pipeline.ts]: typically, when we hardcode a string like this. it is good to add some comments.

Comment (mtarng) [reply src/commands/project/pipeline.ts]: Okay, then you can add the comment
> Pipelines are triggered only by merges into the master branch.

File: src/commands/project/pipeline.ts (hunk 3 above)
Comment (mtarng) [reply src/commands/project/pipeline.ts:105]: Okay, that's fine.

File: src/commands/project/pipeline.ts (hunk 4 above)
Comment (dennisseah) [reply src/commands/project/pipeline.ts]: will do. thanks

File: src/commands/project/pipeline.ts (hunk 3 above)
Comment (dennisseah) [reply src/commands/project/pipeline.ts:105]: thanks

**Review by mtarng — APPROVED** [review]

**Pull request merged**